relay-server/src/utils/bucket_encoding.rs

```diff
@@ -0,0 +1,201 @@
+    pub fn prepare(&self, bucket: &mut Bucket) -> MetricNamespace {
+        let namespace = MetricResourceIdentifier::parse(&bucket.name)
+            .map(|mri| mri.namespace)
+            .unwrap_or(MetricNamespace::Unsupported);
+
+        if let BucketValue::Distribution(ref mut distribution) = bucket.value {
+            let enc = self.global_config.options.metric_bucket_dist_encodings;
+            let enc = enc.for_namespace(namespace);
+
+            if matches!(enc, BucketEncoding::Zstd) {
+                distribution.sort_unstable();
+            }
+        }
+
+        namespace
+    }
+
+    /// Encodes a distribution.
+    pub fn encode_distribution<'data>(
+        &self,
+        namespace: MetricNamespace,
+        dist: &'data [FiniteF64],
+    ) -> io::Result<ArrayEncoding<&'data [FiniteF64]>> {
+        let enc = self.global_config.options.metric_bucket_dist_encodings;
+        let enc = enc.for_namespace(namespace);
+        Self::do_encode(enc, dist)
+    }
+
+    /// Encodes a set.
+    pub fn encode_set<'data>(
+        &self,
+        namespace: MetricNamespace,
+        set: SetView<'data>,
+    ) -> io::Result<ArrayEncoding<SetView<'data>>> {
+        let enc = self.global_config.options.metric_bucket_set_encodings;
+        let enc = enc.for_namespace(namespace);
+        Self::do_encode(enc, set)
+    }
+
+    fn do_encode<T: Encodable>(enc: BucketEncoding, data: T) -> io::Result<ArrayEncoding<T>> {
+        match enc {
+            BucketEncoding::Legacy => Ok(ArrayEncoding::Legacy(data)),
+            BucketEncoding::Array => {
+                Ok(ArrayEncoding::Dynamic(DynamicArrayEncoding::Array { data }))
+            }
+            BucketEncoding::Base64 => base64(data),
+            BucketEncoding::Zstd => zstd(data),
+        }
+    }
+}
+
+/// Dynamic array encoding intended for distribution and set metric buckets.
+#[derive(Clone, Debug, Serialize)]
+#[serde(untagged)]
+pub enum ArrayEncoding<T> {
+    /// The original, legacy, encoding.
+    ///
+    /// Encodes all values as an array of numbers.
+    Legacy(T),
+    /// Dynamic encoding supporting multiple formats.
+    ///
+    /// Adds metadata and adds support for multiple different encodings.
+    Dynamic(DynamicArrayEncoding<T>),
+}
+
+impl<T> ArrayEncoding<T> {
+    /// Name of the encoding.
+    ///
+    /// Should only be used for debugging purposes.
+    pub fn name(&self) -> &'static str {
+        match self {
+            Self::Legacy(_) => "legacy",
+            Self::Dynamic(dynamic) => dynamic.format(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(tag = "format", rename_all = "lowercase")]
+pub enum DynamicArrayEncoding<T> {
+    /// Array encoding.
+    ///
+    /// Encodes all items as an array.
+    Array { data: T },
+    /// Base64 (with padding) encoding.
+    ///
+    /// Converts all items to little endian byte sequences
+    /// and Base64 encodes the raw little endian bytes.
+    Base64 { data: String },
```

Comment:
Do we even need this mode? Even if it seems like a sub-mode of `zstd`, it's still a different case that the consumer needs to be able to support correctly -> more room for errors.